crates/metal-llvm/src/lib.rs

```diff
@@ -1,36 +1,54 @@
 //! Metal library for compiling to LLVM IR using MIR.
 
-use metal_mir::parcel::Module;
+use std::collections::HashMap;
+
+use metal_mir::{
+    parcel::Module,
+    types::{visibility::Visibility, Type},
+};
 
 pub mod expr;
 pub mod primitives;
 pub mod stmt;
 pub mod ty;
 
-use llvm_sys::prelude::{LLVMBuilderRef, LLVMContextRef, LLVMModuleRef, LLVMTypeRef, LLVMValueRef};
+use llvm_sys::{
+    prelude::{LLVMBuilderRef, LLVMContextRef, LLVMModuleRef, LLVMTypeRef, LLVMValueRef},
+    LLVMLinkage,
+};
 
 pub struct LLVMRefs {
     ctx: LLVMContextRef,
     builder: LLVMBuilderRef,
     module: LLVMModuleRef,
+    locals: HashMap<&'static str, LLVMValueRef>,
 }
 
 pub trait CodeGenValue {
-    fn codegen_value(&self, llvm: &LLVMRefs, module: &Module) -> LLVMValueRef;
+    fn codegen_value(&self, llvm: &mut LLVMRefs, module: &Module) -> LLVMValueRef;
 }
 
 pub trait CodeGenType {
     fn codegen_type(&self, llvm: &LLVMRefs, module: &Module) -> LLVMTypeRef;
 }
 
-pub fn get_types(
+pub fn get_types<'a>(
     llvm: &LLVMRefs,
     module: &Module,
-    types: &[metal_mir::types::Type],
+    cap: &usize,
```

Comment:
In the future, we should check if we can add an [`ExactSizeIterator`](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html) bound to the `types` param to remove the need for this one.